### PR TITLE
EventAdmin interface

### PR DESCRIPTION
**Pull request by achristoforides**
# Description

This PR introduces the EventAdmin interface (locally called EM). It follows the OSGi specification for EventAdmin: https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.event.html

# Testing Information

Tests have been added to the `compendium/EM` directory. These tests cover functionality provided by the implementation of `EMEvent`.

### CODE REVIEW

File: compendium/EM/include/cppmicroservices/em/Event.hpp [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+  virtual ~Event() = default;
+
+  /**
+   * @brief Construct a new Event object with the provided topic and event properties.
+   * 
+   * Topics take the following form:
+   *   topic-description := '*' | topic ( '/\*' )?
+   *   topic := token ( '/' token )*
+   *   token := [A-Za-z0-9_.]
```
Comment (carneyweb) [inline compendium/EM/include/cppmicroservices/em/Event.hpp:75]: This comment does not match the implementation in Event.cpp:

  std::string regexExpr = "([A-Za-z0-9_.]+)(\\/[A-Za-z0-9_.]+)*";

"token" needs to have a "+" sign at the end (i.e. token matches 1 or more characters in the set).

File: compendium/EM/test/TestEvent.cpp [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+#include "cppmicroservices/LDAPFilter.h"
+
+#include "gtest/gtest.h"
+
+using cppmicroservices::service::em::Event;
+using cppmicroservices::service::em::EventProperties;
+namespace emc = cppmicroservices::em::Constants;
+
+namespace {
+TEST(EventTest, TestConstructionValidTopic)
```
Comment (carneyweb) [inline compendium/EM/test/TestEvent.cpp:39]: There should be tests that cover topics that are specified with illegal characters that should make sure that the regex being used to validate the topics behaves properly.

**Review by carneyweb — CHANGES_REQUESTED** [review]

File: compendium/EM/test/TestEvent.cpp (hunk 2 above)
Comment (tcormackMW) [reply compendium/EM/test/TestEvent.cpp:39]: Resolved

File: compendium/EM/include/cppmicroservices/em/Event.hpp (hunk 1 above)
Comment (tcormackMW) [reply compendium/EM/include/cppmicroservices/em/Event.hpp]: fixed

File: compendium/EM/include/cppmicroservices/em/Event.hpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+   */
+  bool operator!=(const Event& other) const;
+
+  /**
+   * @brief Returns whether or not the provided property is specified in the event properties.
+   *
+   * @param propName The peropty to search for
+   * @return bool true if the property is defined, false otherwise
+   */
+  bool ContainsProperty(const std::string& propName) const;
```
[the remark this answers is not in the document]
Comment (tcormackMW) [reply compendium/EM/include/cppmicroservices/em/Event.hpp:115]: I disagree

File: compendium/EM/include/cppmicroservices/em/EventAdmin.hpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    {
+      public:
+        virtual ~EventAdmin() = default;
+        EventAdmin(EventAdmin&&) = default;
+        EventAdmin(EventAdmin&) = default;
+        EventAdmin& operator=(EventAdmin const&) = default;
+        EventAdmin& operator=(EventAdmin&&) = default;
+        /**
+         * @brief Initiates the asynchronous delivery of an event. If the \c DELIVERY_ASYNC_UNORDERED
+         * event property is specified, it will not guaranteed that the event will be delivered in the
```
Comment (jeffdiclemente) [inline compendium/EM/include/cppmicroservices/em/EventAdmin.hpp:58]: minor grammar: "it will not guaranteed" => "it will not be guaranteed"

**Review by jeffdiclemente — CHANGES_REQUESTED** [review]
is the third_party/boost/nowide update necessary for this change?

**Comment by mingweiLIU** [comment]
Is this feature ready?

**Comment by tcormackMW** [comment]
> is the third_party/boost/nowide update necessary for this change?

@jeffdiclemente I don't think so. I am not sure how those got into the changelist

File: compendium/EM/include/cppmicroservices/em/EMConstants.hpp [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+    /**
+     * Event property identifying the symbolic name of the bundle where the event was sent from. The
+     * value of this property is of type <code>std::string</code>.
+     */
+    US_em_EXPORT extern std::string const BUNDLE_SYMBOLICNAME;
+
+    /**
+     * Event property identifying the version of the bundle where the event was sent from. The value of
+     * this property is of type <code>std::string</code>.
```
[the remark this answers is not in the document]
Comment (tcormackMW) [reply compendium/EM/include/cppmicroservices/em/EMConstants.hpp:54]: I agree

File: compendium/EM/include/cppmicroservices/em/EMConstants.hpp [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+    /**
+     * Event property containing a service reference of your choosing. The value of this property is of
+     * type <code>cppmicroservices::ServiceReferenceU</code>.
+     */
+    US_em_EXPORT extern std::string const SERVICE;
+
+    /**
+     * Event property containing the id of the provided service reference in \c SERVICE . The value of
+     * this property is of type <code>unsigned int</code>.
```
[the remark this answers is not in the document]
Comment (tcormackMW) [reply compendium/EM/include/cppmicroservices/em/EMConstants.hpp:161]: I agree

File: compendium/EM/include/cppmicroservices/em/EMConstants.hpp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+     *
+     * The value of this property is of type <code>std::vector<std::string></code>
+     */
+    US_em_EXPORT extern std::string const EVENT_TOPIC;
+
+    /**
+     * Event property containing the exception object if one was thrown. The value of this property is
+     * of type equal to that of the string representation of \c EXCEPTION_CLASS .
+     */
+    US_em_EXPORT extern std::string const EXCEPTION;
```
[the remark this answers is not in the document]
Comment (tcormackMW) [reply compendium/EM/include/cppmicroservices/em/EMConstants.hpp:132]: I agree

File: compendium/EM/include/cppmicroservices/em/EventAdmin.hpp (hunk 4 above)
Comment (tcormackMW) [reply compendium/EM/include/cppmicroservices/em/EventAdmin.hpp:58]: Fixed

**Comment by jeffdiclemente** [comment]
> Is this feature ready?

not yet. what we have in this PR is the service interface. The implementation of the interface still needs to be done.

**Review by jeffdiclemente — APPROVED** [review]

**Pull request merged**